datasets/stream_spec.py

```diff
@@ -0,0 +1,192 @@
+        """
+        classes_list = list(range(self.n_classes))
+        random_state = np.random.RandomState(self.random_seed)
+
+        task_classes = list()
+        for _ in range(self.n_tasks):
+            if len(classes_list) < self.n_slots:
+                classes_list.extend(range(self.n_classes))
+
+            selected_classes = [None]
+            label = None
+            for _ in range(self.n_slots):
+                while label in selected_classes:
+                    label = random_state.choice(classes_list, replace=False)
+                selected_classes.append(label)
+
+            selected_classes.remove(None)
+
+            for c in selected_classes:
+                classes_list.remove(c)
+            task_classes.append(list(selected_classes))
+
+        # assign remaining classes to the first task
+        for c in classes_list:
+            if c not in task_classes[0]:
+                task_classes[0].append(c)
+
+        class_mapping = dict()
+        last_class = 0
+        for task_class in task_classes:
+            for c in task_class:
+                if c not in class_mapping:
+                    class_mapping[c] = last_class
+                    last_class += 1
+
+        new_task_classes = list()
+        for task_class in task_classes:
+            new_classes = sorted(class_mapping[c] for c in task_class)
+            new_task_classes.append(new_classes)
+        return new_task_classes
+
+    def __next__(self):
+        while self.current_task < self.n_tasks:
+            yield self.task_classes[self.current_task]
+            self.current_task += 1
+
+    def __iter__(self):
+        return next(self)
+
+    @property
+    def drifted_classes(self):
+        """return all classes for which drift has occured so far"""
+        seen_classes = set()
+        drifted_cl = set()
+        for i in range(min(self.current_task+1, self.n_tasks)):
+            for c in self.task_classes[i]:
+                if c in seen_classes:
+                    drifted_cl.add(c)
+                else:
+                    seen_classes.add(c)
+        return list(drifted_cl)
+
+    @property
+    def drifted_classes_last_task(self):
+        """return classes from last task for which dirft has occured"""
+        current_classes = self.task_classes[self.current_task]
+        current_classes = set(current_classes)
+        drifted_classes = set(self.drifted_classes)
+        current_drifted = set.intersection(current_classes, drifted_classes)
+        return list(current_drifted)
+
+    @property
+    def new_classes_last_task(self):
+        """return classes from current task"""
+        current_classes = self.task_classes[self.current_task]
+        current_classes = set(current_classes)
+        drifted_classes = set(self.drifted_classes)
+        current_drifted = current_classes - drifted_classes
+        return list(current_drifted)
+
```

Comment:
The method name `last_task` is a bit confusing since we are returning drifted classes from the current task.

Reply:
Totaly agree on this one.
I propose to change names of properties in following way:
drifted_classes -> all_drifted_classes
drifted_classes_last_task -> drifted_classes
new_classes_last_task -> new_classes
I think this should work better. Let me know what you think and I will change it.

Reply:
Yes, I think the new proposed names are more meaningful. You can go ahead and change them.